executor/aggfuncs/func_count_distinct.go

```diff
@@ -0,0 +1,460 @@
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/types/json"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/set"
+)
+
+type partialResult4CountDistinctInt struct {
+	partialResult4Count
+	valSet set.Int64Set
+}
+
+type countOriginalWithDistinct4Int struct {
+	baseCount
+}
+
+func (e *countOriginalWithDistinct4Int) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4CountDistinctInt{
+		partialResult4Count: 0,
+		valSet:              set.NewInt64Set(),
+	})
+}
+
+func (e *countOriginalWithDistinct4Int) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4CountDistinctInt)(pr)
+	p.partialResult4Count = 0
+	p.valSet = set.NewInt64Set()
+}
+
+func (e *countOriginalWithDistinct4Int) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4CountDistinctInt)(pr)
+	chk.AppendInt64(e.ordinal, p.partialResult4Count)
+	return nil
+}
+
+func (e *countOriginalWithDistinct4Int) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4CountDistinctInt)(pr)
+
+	for _, row := range rowsInGroup {
+		input, isNull, err := e.args[0].EvalInt(sctx, row)
+		if err != nil {
+			return err
+		}
+		if isNull {
+			continue
+		}
+		if p.valSet.Exist(input) {
+			continue
+		}
+		p.valSet.Insert(input)
+
+		p.partialResult4Count++
+	}
+
+	return nil
+}
+
+type partialResult4CountDistinctReal struct {
+	partialResult4Count
+	valSet set.Float64Set
+}
+
+type countOriginalWithDistinct4Real struct {
+	baseCount
+}
+
+func (e *countOriginalWithDistinct4Real) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4CountDistinctReal{
+		partialResult4Count: 0,
+		valSet:              set.NewFloat64Set(),
+	})
+}
+
+func (e *countOriginalWithDistinct4Real) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4CountDistinctReal)(pr)
+	p.partialResult4Count = 0
+	p.valSet = set.NewFloat64Set()
+}
+
+func (e *countOriginalWithDistinct4Real) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4CountDistinctReal)(pr)
+	chk.AppendInt64(e.ordinal, p.partialResult4Count)
+	return nil
+}
+
+func (e *countOriginalWithDistinct4Real) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4CountDistinctReal)(pr)
+
+	for _, row := range rowsInGroup {
+		input, isNull, err := e.args[0].EvalReal(sctx, row)
+		if err != nil {
+			return err
+		}
+		if isNull {
+			continue
+		}
+		if p.valSet.Exist(input) {
+			continue
+		}
+		p.valSet.Insert(input)
+
+		p.partialResult4Count++
+	}
+
+	return nil
+}
+
+type partialResult4CountDistinctDecimal struct {
+	partialResult4Count
+	valSet set.StringSet
```

Comment:
Should we also update avgDistinct and sumDistinct to use StringSet?

Reply:
They are already using `StringSet`.